Compilation error when using `unreachable` macro

<!--
If this is a security issue, please report it following the
[security reporting procedure](https://github.com/model-checking/kani/security/policy).
-->

<!--
Thank you for filing a bug report! 🐛
Please provide a short summary of the issue, along with the information necessary to replicate.
-->

I tried this code:
<!--
If the reproducing code is small, please post it here.
Otherwise, please attach the relevant files to this issue.
Alternatively, post a link to the repository and branch that exposes the issue.
-->
```rust
#[kani::proof]
fn main() {
    let x = 5;
    unreachable!("The value of x is {}", x);
}
```

using the following command line invocation:

```
kani test.rs
```

with Kani version: cb1a6068a15

Compilation fails with:
```
$ kani test.rs 
error[E0061]: this function takes 1 argument but 2 arguments were supplied
   --> test.rs:4:5
    |
4   |     unreachable!("The value of x is {}", x);
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ argument of type `&'static str` unexpected
    |
note: function defined here
   --> /home/ubuntu/git/kani/library/kani/src/lib.rs:124:8
    |
124 | pub fn panic(message: &'static str) -> ! {
    |        ^^^^^
    = note: this error originates in the macro `unreachable` (in Nightly builds, run with -Z macro-backtrace for more info)
help: remove the extra argument
   --> |/home/ubuntu/git/kani/library/std/src/lib.rs:147:9
    |
147 |         kani::panic(unreachable!("The value of x is {}", x)));
    |         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

error: aborting due to previous error

For more information about this error, try `rustc --explain E0061`.
Error: "Failed to compile crate."
Error: /home/ubuntu/git/kani/target/debug/kani-compiler exited with status exit status: 1
```

--- a/library/std/src/lib.rs
+++ b/library/std/src/lib.rs
@@ -143,8 +143,14 @@ macro_rules! unreachable {
     ($($msg:literal)? $(,)?) => (
         kani::panic(concat!("internal error: entered unreachable code: ", $($msg)?))
     );
+    // Needed for 2018 and older rust editions.
+    // TODO: Is it possible to trigger an error if 2021 and above?
+    ($($msg:expr)? $(,)?) => (
+        kani::panic(concat!("internal error: entered unreachable code: ", stringify!($($msg)?)))
+    );
     ($fmt:expr, $($arg:tt)*) => (
-        kani::panic(concat!("internal error: entered unreachable code: ", $fmt), stringify!($($arg)*)));
+        kani::panic(concat!("internal error: entered unreachable code: ",
+        stringify!($fmt, $($arg)*))));
 }
 
 #[macro_export]
@@ -155,6 +161,8 @@ macro_rules! panic {
     ($msg:literal $(,)?) => ({
         kani::panic(concat!($msg));
     });
+    // Needed for 2018 and older rust editions.
+    // TODO: Is it possible to trigger an error if 2021 and above?
     ($msg:expr $(,)?) => ({
         kani::panic(stringify!($msg));
     });

--- a/library/std/src/lib.rs
+++ b/library/std/src/lib.rs
@@ -140,33 +140,54 @@ macro_rules! eprintln {
 
 #[macro_export]
 macro_rules! unreachable {
+    // The argument, if present, is a literal that represents the error message, i.e.:
+    // `unreachable!("Error message")` or `unreachable!()`
     ($($msg:literal)? $(,)?) => (
         kani::panic(concat!("internal error: entered unreachable code: ", $($msg)?))
     );
-    // Needed for 2018 and older rust editions.
+    // The argument is an expression, such as a variable.
+    // ```
+    // let msg = format!("Error: {}", code);
+    // unreachable!(msg);
+    // ```
+    // This was supported for 2018 and older rust editions.
     // TODO: Is it possible to trigger an error if 2021 and above?
+    // https://github.com/model-checking/kani/issues/1375
     ($($msg:expr)? $(,)?) => (
         kani::panic(concat!("internal error: entered unreachable code: ", stringify!($($msg)?)))
     );
-    ($fmt:expr, $($arg:tt)*) => (
+    // The first argument is the format and the rest contains tokens to be included in the msg.
+    // `unreachable!("Error: {}", code);`
+    ($fmt:literal, $($arg:tt)*) => (
         kani::panic(concat!("internal error: entered unreachable code: ",
         stringify!($fmt, $($arg)*))));
 }
 
 #[macro_export]
 macro_rules! panic {
+    // No argument is given.
     () => (
         kani::panic("explicit panic")
     );
+    // The argument is a literal that represents the error message, i.e.:
+    // `panic!("Error message")`
     ($msg:literal $(,)?) => ({
         kani::panic(concat!($msg));
     });
-    // Needed for 2018 and older rust editions.
+    // The argument is an expression, such as a variable.
+    // ```
+    // let msg = format!("Error: {}", code);
+    // panic!(msg);
+    // ```
+    // This was supported for 2018 and older rust editions.
     // TODO: Is it possible to trigger an error if 2021 and above?
+    // https://github.com/model-checking/kani/issues/1375
     ($msg:expr $(,)?) => ({
         kani::panic(stringify!($msg));
     });
-    ($msg:expr, $($arg:tt)*) => ({
-        kani::panic(stringify!($msg, $($arg)*));
+    // The first argument is the format and the rest contains tokens to be included in the msg.
+    // `panic!("Error: {}", code);`
+    ($msg:literal, $($arg:tt)+) => ({
+        kani::panic(stringify!($msg, $($arg)+));
     });
 }

--- a/tests/expected/unreachable/expected
+++ b/tests/expected/unreachable/expected
@@ -1,0 +1,6 @@
+Failed Checks: internal error: entered unreachable code: msg
+Failed Checks: internal error: entered unreachable code: 
+Failed Checks: internal error: entered unreachable code: Error message
+Failed Checks: internal error: entered unreachable code: "Unreachable message with arg {}", "str"
+Failed Checks: internal error: entered unreachable code: "{}", msg
+

--- a/tests/expected/unreachable/expected
+++ b/tests/expected/unreachable/expected
@@ -1,5 +1,5 @@
 Failed Checks: internal error: entered unreachable code: msg
-Failed Checks: internal error: entered unreachable code: 
+Failed Checks: internal error: entered unreachable code:
 Failed Checks: internal error: entered unreachable code: Error message
 Failed Checks: internal error: entered unreachable code: "Unreachable message with arg {}", "str"
 Failed Checks: internal error: entered unreachable code: "{}", msg

--- a/tests/expected/unreachable/unreach_msg.rs
+++ b/tests/expected/unreachable/unreach_msg.rs
@@ -1,0 +1,15 @@
+// Copyright Kani Contributors
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+
+//! Checks that our macro override supports different types of messages.
+#[kani::proof]
+fn check_unreachable() {
+    let msg = "Oops.";
+    match kani::any::<u8>() {
+        0 => unreachable!(),
+        1 => unreachable!("Error message"),
+        2 => unreachable!("Unreachable message with arg {}", "str"),
+        3 => unreachable!("{}", msg),
+        _ => unreachable!(msg),
+    }
+}

